### PR TITLE
Add original attachments when forwarding a mail

### PR DESCRIPTION
**Pull request by StCyr**
Fix to https://github.com/nextcloud/mail/issues/246

Please review

Signed-off-by: Cyrille Bollu <cyrpub@bollu.be>

### CODE REVIEW

**Comment by ChristophWurst** [comment]
I don't see why we'd want this for replies. IMO this only makes sense for forwards :thinking:

**Review by ChristophWurst — CHANGES_REQUESTED** [review]
Conceptually I would do it a bit differently.

Currently you have a dedicated back-end route that copies the contents of an existing message into the temporary attachments storage, so later it can be used like a file that was just uploaded by the user.

We currently already have two types of attachments: uploaded ones and selections from the Nextcloud file system. I'd extend this to a third type that represents an original message or attachments of such a message. Which means there is no dedicated logic to duplicate any existing data to fake any uploaded file, but we'd extend the sending logic to fetch other data if necessary. IMO this would be less error-prone and also more efficient as we'd only fetch the original data if the user selects to include it.

**Comment by StCyr** [comment]
Ok, I think I understand your idea

**Comment by StCyr** [comment]
Hello @ChristophWurst ,

I think this updated PR better implements your idea.

Could you please have a new look at it?

**Comment by StCyr** [comment]
If it's the good way to go I still have to:

1- Update the integration tests
2- see if I can get a better mime type for the attachments (currently "application/octet-stream")

**Comment by StCyr** [comment]
Please review @ChristophWurst

File: src/components/ComposerAttachments.vue [hunk 1]
```diff
@@ -98,7 +98,7 @@ export default {
 				fileName: name,
 				displayName: trimStart('/')(name),
 				id,
-				isLocal: id !== undefined,
+				type: id !== undefined ? 'local' : 'cloud',
```
Comment (ChristophWurst) [inline src/components/ComposerAttachments.vue:101]: this change will conflict with https://github.com/nextcloud/mail/pull/3998. The PR that is merged later has to be adjusted

**Comment by ChristophWurst** [comment]
> Please review @ChristophWurst

It looks a lot better conceptually :)

The only thing I still wonder if we can't create the new attachments from downloading the old attachments directly. Like instead of storing them locally to a temporary file and creating a database entry. I think that should be possible. Then the mime would be 1:1 the same.

**Comment by StCyr** [comment]
> The only thing I still wonder if we can't create the new attachments from downloading the old attachments directly. Like instead of storing them locally to a temporary file and creating a database entry. I think that should be possible. Then the mime would be 1:1 the same.

I don't know any other method to create a `ISimpleFile` 😁

**Comment by ChristophWurst** [comment]
> I don't know any other method to create a `ISimpleFile`

I'll try to dive into this soon :)

**Review by st3iny — APPROVED** [review]
LGTM! Tested locally and it works fine.

**Comment by StCyr** [comment]
> 
> 
> LGTM! Tested locally and it works fine.

Glad you like it. But, @ChristophWurst has a remark that it might be better if we could just the "temporary file creation" step, and I quite agree with him: This would be more elegant

I'm wondering though if, in such case, the new attachment wouldn't get deleted from the persistent storage if the original email (with the original attachment) is deleted.... ❓

**Comment by st3iny** [comment]
Oops, I thought you wanted to fix this in a followup PR.

What about adding a method to AttachmentService that takes an attachment instead of an UploadedFile and then also adjust AttachmentStorage to save a string buffer?

```php
// Mockup for MailTransmission.handleEmailAttachment()
$att = $folder->getAttachment($attachmentMessage->getUid(), $attachment['id'])
$newAtt = $this->attachmentService->copyAttachment($userId, $att)

// Mockup for AttachmentStorage.saveContents()
$file->putContent($att->getContents())
```

**Comment by StCyr** [comment]
works good now but mimeType is set to 'application/octet-stream'

**Comment by StCyr** [comment]
It looks like all tests are green :-)

**Comment by StCyr** [comment]
Hey @ChristophWurst, I've added a unit test for this new feature. Could you please have a look? That's the first unit test I've ever writen.

File: lib/Model/Message.php [hunk 2]
```diff
@@ -222,6 +223,25 @@ public function getLocalAttachments(): array {
...
+	 *
+	 */
+	public function addAttachmentFromAttachment(Attachment $attachment) {
+		$part = new Horde_Mime_Part();
+		$part->setCharset('us-ascii');
+		$part->setDisposition('attachment');
+		$part->setName($attachment->getName());
+		$part->setContents($attachment->getContents());
+		$part->setType($attachment->getType());
+		$this->cloudAttachments[] = $part;
```
Comment (ChristophWurst) [inline lib/Model/Message.php:242]: :thinking: my IDE says `$this->cloudAttachments` is only read via `\OCA\Mail\Model\Message::getCloudAttachments` and that method is only used in tests.

File: lib/Model/Message.php [hunk 3]
```diff
@@ -222,6 +223,25 @@ public function getLocalAttachments(): array {
...
+	 *
+	 * @return void
+	 *
+	 * Adds a file that's coming from another email's attachment (typical
+	 * use case is forwarding a message)
+	 *
+	 */
+	public function addAttachmentFromAttachment(Attachment $attachment) {
+		$part = new Horde_Mime_Part();
+		$part->setCharset('us-ascii');
```
Comment (ChristophWurst) [inline lib/Model/Message.php:237]: this sounds like a assumption. do we know if this is always correct?

File: lib/Service/MailTransmission.php [hunk 4]
```diff
@@ -321,6 +337,29 @@ private function handleLocalAttachment(string $userId, array $attachment, IMessa
...
+	 *
+	 */
+	private function handleEmailAttachment(string $userId, array $attachment, IMessage $message) {
+
+		// Gets attachment from other message
+		$attachmentMessage = $this->mailManager->getMessage($userId, (int)$attachment['messageId']);
+		$mailbox = $this->mailManager->getMailbox($userId, $attachmentMessage->getMailboxId());
+		$account = $this->accountService->find($userId, $mailbox->getAccountId());
+		$folder = $account->getMailbox($mailbox->getName());
+		$attachment = $folder->getAttachment($attachmentMessage->getUid(), $attachment['id']);
```
Comment (ChristophWurst) [inline lib/Service/MailTransmission.php:357]: I'm having mixed feelings here

the "folder" code is the type of code I really want to get rid of. Can we somehow use `\OCA\Mail\IMAP\MessageMapper::getRawAttachments` or add a similar method instead? It should do the as `$folder->getAttachment` but with the new (testable) code architecture

File: tests/Integration/Service/MailTransmissionIntegrationTest.php [hunk 5]
```diff
@@ -159,6 +163,22 @@ public function testSendMailWithCloudAttachment() {
 		$this->addToAssertionCount(1);
 	}
 
+	/* TODO
+	public function testSendMailWithAttachmentFromAttachment() {
```
Comment (ChristophWurst) [inline tests/Integration/Service/MailTransmissionIntegrationTest.php:167]: dead? why is this commented out?

Comment (StCyr) [reply tests/Integration/Service/MailTransmissionIntegrationTest.php:167]: didn't implement it yet (don't know if I will though)

File: lib/Service/MailTransmission.php (hunk 4 above)
Comment (StCyr) [reply lib/Service/MailTransmission.php:357]: of course. I will do this

File: tests/Integration/Service/MailTransmissionIntegrationTest.php (hunk 5 above)
Comment (ChristophWurst) [reply tests/Integration/Service/MailTransmissionIntegrationTest.php:167]: then just drop that code and we add the test when the actual feature is added :)

File: lib/Model/Message.php (hunk 3 above)
Comment (StCyr) [reply lib/Model/Message.php:237]: Well, I've just copied if from addAttachmentFromFiles and addLocalAttachment 🙈

Comment (ChristophWurst) [reply lib/Model/Message.php:237]: Well then yolo on :speak_no_evil:

File: lib/Service/MailTransmission.php (hunk 4 above)
Comment (StCyr) [reply lib/Service/MailTransmission.php]: should be good now

File: lib/Model/IMessage.php [hunk 6]
```diff
@@ -130,6 +130,12 @@ public function getCloudAttachments(): array;
 	 */
 	public function getLocalAttachments(): array;
 
+	/**
+	 * @param string $name
+	 * @param string $content
+	 */
+	public function addAttachmentFromAttachment(string $name, string $content);
```
Comment (ChristophWurst) [inline lib/Model/IMessage.php:137]:
```suggestion
	public function forwardAttachment(string $name, string $content);
```

:speak_no_evil:

Comment (ChristophWurst) [reply lib/Model/IMessage.php:137]: and return type can be specified as `void`

**Comment by ChristophWurst** [comment]
I'll review/integrate this soon. The part with the cloud attachments I still don't get. Not specifically your code, but also the current one. It must be some dark magic that I don't comprehend.

File: lib/Model/IMessage.php (hunk 6 above)
Comment (StCyr) [reply lib/Model/IMessage.php:137]: `forwardAttachment` is better, but wouldn't ´addForwardedAttachment´ betterer?

File: lib/Model/Message.php (hunk 2 above)
Comment (StCyr) [reply lib/Model/Message.php:244]: I don't understand your comment. But, I must say I don't really understand this cloudAttachments vs localAttachments distinction: It doesn't seem to make much sense

File: lib/Model/IMessage.php (hunk 6 above)
Comment (ChristophWurst) [reply lib/Model/IMessage.php:137]: yes :)

Comment (ChristophWurst) [reply lib/Model/IMessage.php:137]: it's definitely the betterestest option

Comment (StCyr) [reply lib/Model/IMessage.php:137]: 🤣

File: src/components/ComposerAttachments.vue (hunk 1 above)
Comment (ChristophWurst) [reply src/components/ComposerAttachments.vue:101]: done

**Review by ChristophWurst — APPROVED** [review]
Rebased to resolve the conflict with my other PR and cleaned up a bit of technical dept with the attachments handling. We could simplify the code quite a bit actually. And I could optimize the IMAP message mapper to even only download the parts that we are interested in, the previous version fetched the all parts and did the filtering afterwards.

Really really really good job. Thanks a lot @StCyr!

File: lib/Model/Message.php (hunk 2 above)
Comment (ChristophWurst) [reply lib/Model/Message.php]: https://github.com/nextcloud/mail/pull/3950#pullrequestreview-553717905 I was wrong :wink:

**Pull request merged**